File: array/sorting/0283_move_zeros.js
/*
Given an integer array nums, move all 0's to the end of it while maintaining 
the relative order of the non-zero elements.

Note that you must do this in-place without making a copy of the array.
*/
// https://leetcode.com/problems/move-zeroes/

// two pointers
// time complexity O(n)
// space complexity O(1)
var moveZeroes = function (nums) {
  let left = 0;
  let right = nums.length - 1;
  while (left < right) {
    if (nums[left] !== 0) {
      left++;
    } else if (nums[right] === 0) {
      right--;
    } else {
      for (let i = left; i < right; i++) {
        [nums[i], nums[i + 1]] = [nums[i + 1], nums[i]];
      }
      right--;
    }
  }
  return nums;
};

console.log(moveZeroes([1, 2, 3, 0, 4, 5]));
console.log(moveZeroes([0, 1, 0, 3, 12]));
console.log(moveZeroes([0, 0, 1]));

File: array/sorting/0283_move_zeros.py
# Given an integer array nums, move all 0's to the end of it while maintaining 
# the relative order of the non-zero elements.

# Note that you must do this in-place without making a copy of the array.

# https://leetcode.com/problems/move-zeroes/

# stable in-place sort (bubble sort)
# time complexity O(n^2)
# space complexity O(1)
class Solution:
    def moveZeroes(self, nums: list[int]) -> None:
        l = len(nums)
        for i in range(l-1, -1, -1):
            if nums[i] == 0:
                j = i
                while j < l-1:
                    nums[j], nums[j+1] = nums[j+1], nums[j]
                    j += 1

# optimized but still with O(n^2) time complexity
class Solution:
    def moveZeroes(self, nums: list[int]) -> None:
        l = len(nums)
        left, right = 0, l - 1
        while left < right:
            if nums[left] != 0:
                left += 1
            elif nums[right] == 0:
                right -= 1
            else:
                for i in range(left, right):
                    nums[i], nums[i + 1] = nums[i + 1], nums[i]
                right -= 1
          
            